prov/gni/src/gnix_mr.c

```diff
@@ -771,127 +973,292 @@ int _gnix_mr_cache_flush(gnix_mr_cache_t *cache)
+	if (!dlist_empty(&tmp)) {
+		dlist_for_each_safe(&tmp, found_entry, next_entry, siblings) {
+			dlist_remove(&found_entry->siblings);
+			dlist_insert_tail(&found_entry->siblings,
+					&(*entry)->children);
+			if (!dlist_empty(&found_entry->children)) {
+				/* move the entry's children to the sibling tree
+				 * and decrement the reference count */
+				dlist_splice_tail(&(*entry)->children,
+						&found_entry->children);
+				__mr_cache_entry_put(cache, found_entry);
+			}
+		}
+		assert(dlist_empty(&tmp));
+		__mr_cache_entry_get(cache, *entry);
+	}
+
+	cache->misses++;
 
 	return ret;
 }
 
-static int __mr_cache_lookup_stale_fastpath(
-		gnix_mr_cache_t *cache,
-		gnix_mr_cache_key_t *key,
-		gnix_mr_cache_entry_t **entry)
+static int __mr_cache_search_stale(
+		gnix_mr_cache_t          *cache,
+		struct gnix_fid_domain   *domain,
+		uint64_t                 address,
+		uint64_t                 length,
+		gni_cq_handle_t          dst_cq_hndl,
+		uint32_t                 flags,
+		uint32_t                 vmdh_index,
+		gnix_mr_cache_entry_t    **entry,
+		gnix_mr_cache_key_t      *key)
 {
+	int ret;
 	RbtStatus rc;
 	RbtIterator iter;
-	int ret = -FI_ENOENT;
-	gnix_mr_cache_key_t *e_key;
-	gnix_mr_cache_entry_t *current_entry;
+	gnix_mr_cache_key_t *mr_key;
+	gnix_mr_cache_entry_t *mr_entry, *tmp;
 
-	/* initialize to NULL */
-	*entry = NULL;
+	GNIX_INFO(FI_LOG_MR, "searching for stale entry, key=%llu:%llu\n",
+			key->address, key->length);
 
-	iter = rbtFind(cache->stale, key);
-	if (iter) {
-		rbtKeyValue(cache->stale, iter, (void **) &e_key,
-				(void **) &current_entry);
+	iter = rbtFindLeftmost(cache->stale.rb_tree, (void *) key,
+			__find_overlapping_addr);
+	if (!iter)
+		return -FI_ENOENT;
 
-		atomic_set(&current_entry->ref_cnt, 1);
+	rbtKeyValue(cache->stale.rb_tree, iter, (void **) &mr_key,
+			(void **) &mr_entry);
 
-		/* clear the element from the stale cache */
-		rbtErase(cache->stale, iter);
-		atomic_dec(&cache->stale_elements);
+	GNIX_INFO(FI_LOG_MR,
+			"found a matching entry, found=%llu:%llu key=%llu:%llu\n",
+			mr_key->address, mr_key->length,
+			key->address, key->length);
 
-		dlist_remove(&current_entry->lru_entry);
 
-		GNIX_INFO(FI_LOG_MR,
-				"moving key %llu:%llu from stale into inuse\n",
-				current_entry->key.address,
-				current_entry->key.length);
-
-		rc = rbtInsert(cache->inuse, &current_entry->key,
-				current_entry);
-		if (rc != RBT_STATUS_OK) {
-			GNIX_WARN(FI_LOG_MR, "failed to insert stale entry "
-					"into inuse cache, rc=%d", rc);
-
-			__mr_cache_entry_destroy(current_entry);
-			*entry = NULL;
-			ret = FI_ENOSPC;
+	/* if the entry that we've found completely subsumes
+	 * the requested entry, just return a reference to
+	 * that existing registration
+	 */
+	if (__can_subsume(mr_key, key)) {
+		ret = __mr_cache_search_inuse(cache, domain, address,
+				length, dst_cq_hndl, flags, vmdh_index,
+				&tmp, mr_key);
```

Comment:
We need to search the inuse tree again because the new element pulled from
the stale list might be adjacent to a list we didn't find in the inuse
tree during the first pass.

Consider the following case:

Reg: addr=4096 len=8192
Dereg: addr=4096 len=8192 (moved to stale)
Reg: addr=12288 len=8192
Dereg: addr=12288 len=8192 (moved to stale)
Reg: addr=4096 len=8192 (moved from stale to inuse)
Reg: addr=16384 len=4096 (move addr=12288 len=8192 from stale to inuse)

Now we have two entries in the tree that are adjacent. This invalidates
some of the assumptions that are being made, so the second search through
the inuse tree would find the adjacent entry to the entry being removed
from stale. The two (or more) entries would be merged at this point.
-- Jim




On 1/6/16, 1:04 PM, "Zach Tiffany" <notifications@github.com> wrote:

>In 
>prov/gni/src/gnix_mr.c
><https://github.com/ofi-cray/libfabric-cray/pull/582#discussion_r48994937>
>:
>> -				current_entry);
>> -		if (rc != RBT_STATUS_OK) {
>> -			GNIX_WARN(FI_LOG_MR, "failed to insert stale entry "
>> -					"into inuse cache, rc=%d", rc);
>> -
>> -			__mr_cache_entry_destroy(current_entry);
>> -			*entry = NULL;
>> -			ret = FI_ENOSPC;
>> +	/* if the entry that we've found completely subsumes
>> +	 * the requested entry, just return a reference to
>> +	 * that existing registration
>> +	 */
>> +	if (__can_subsume(mr_key, key)) {
>> +		ret = __mr_cache_search_inuse(cache, domain, address,
>> +				length, dst_cq_hndl, flags, vmdh_index,
>> +				&tmp, mr_key);
>Why do you need to "search_inuse" again here? __mr_cache_register() does
>"search_inuse" before calling into "search_stale" to get us to this
>point. Doesn't that meant we're guaranteed to not have a matching in-use
>element?
>—
>Reply to this email directly or
>view it on GitHub 
><https://github.com/ofi-cray/libfabric-cray/pull/582/files#r48994937>.
>
>